Lane Connector tool - Quick setup

kianzarrin opened the pull request:
fixes #703
fixes #617 
see https://github.com/CitiesSkylinesMods/TMPE/pull/706#issuecomment-591700790 for pictures

### Scope
- Equal junctions are excluded.
- 4 way junctions  EDIT: are supported
- ~~only connects lane arrows for one half of the road.~~ connects lane arrows for both halfs
- lane connections do not overlap. EDIT: its more like a stay in lane for T junctions.
- ~~lane arithmetic should work.~~
- guide message is shown when nothing happens.
- changed shortcut key to `CTRL+S` which conflicts with mod tools.
~~stay in lane no longer does anything if the junction lanes do not match. Is that OK?~~
- some times user wants lane connections to criss-cross but that is outside of the scope of this review.
EDIT: - stay in lane does not connect if there is only one lane: #617
### Notes:
- I removed the stay in lane code because my code is a more general form of stay in lane.
- the only functional change to stay in lane is when lane arithmetic does not match, outer lanes are connected first rather than inner lanes. I consider this to be improvement.
- Tram tracks are outside the scope of this review. Tram tracks already are broken for stay in lane see #705. EDIT: but I added simple `CanConnect()` Condition for tracks which mostly solves #705. Asymmetrical tracks and mixed car/tram roads are still unsupported and out of the scope of this review. see https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/issues/705#issuecomment-586749537 for pictures
- EDIT: I fixed a problem with stay in lane when stay in lane : when `stayInLaneMode == None` selected node markers disappeared. 

### Crowdin
I modified the crowding Hot-Key : https://crowdin.com/translate/tmpe/22/en-en#8606 . 

EDIT: added guide crowdin keys:
```
"TMPE_GUIDE_HEAD_LaneConnectorTool:stay-in-lane is not supported for this setup"=>"Stay in lane"
"TMPE_GUIDE_BODY_LaneConnectorTool::stay-in-lane  is not supported for this setup"=>"Stay-in-lane is not supported for this setup. Make sure  are more than 1 lanes in each direction . Stay-in-lane connects lanes such that they do not crisscross.

aubergine10 commented:
This looked odd behaviour (note: Left hand traffic) - it's two asymmetric roads, facing opposite directions, with a 1-way road exiting the bottom of the junction:

![image](https://user-images.githubusercontent.com/1386719/74615248-ff258200-5116-11ea-89cb-72dba8e985d1.png)

What I would expect:

![image](https://user-images.githubusercontent.com/1386719/74615264-1bc1ba00-5117-11ea-8a49-d8ba7cc20b27.png)

Also, still seems to be using `Shift + S` rather than `Ctrl + S`?

aubergine10 commented:
Another odd one: Road on left = 2 lane 2 way; Road on right = asymmetric, Road at bottom = 2 lane 2 way with bike lanes.

![image](https://user-images.githubusercontent.com/1386719/74615286-4a3f9500-5117-11ea-98da-7fa19bc32627.png)

aubergine10 commented:
Asymmetric roads at bottom, with a 2-way road at the top: 
![image](https://user-images.githubusercontent.com/1386719/74615362-f7b2a880-5117-11ea-8e57-43ac03e13414.png)

aubergine10 commented:
4-lane 2-way medim road at bottom. One way exit on left, asymmetric straight on, 1-way entry on right:

![image](https://user-images.githubusercontent.com/1386719/74615403-511ad780-5118-11ea-9ee8-6b29a37b8f08.png)

2-lane 1-way exiting top, 2-lane 1-way entering right, 4-lane 2-way medium road at bottom:

![image](https://user-images.githubusercontent.com/1386719/74615438-9fc87180-5118-11ea-814c-825a100032a9.png)

aubergine10 commented:
All 2-lane 1-way streets: Entry from bottom, other two are exits:

![image](https://user-images.githubusercontent.com/1386719/74615458-db633b80-5118-11ea-9405-32a5e8fba657.png)

aubergine10 commented:
Medium road (4 lane, 2 way) entering a roundabout consisting of 2-lane 1-way roads:

![image](https://user-images.githubusercontent.com/1386719/74615475-136a7e80-5119-11ea-90c6-b4cc8384528f.png)

Maybe repeat applications of the quick setup could add more "straight on" routes, on first application you get image above, second application would add one straight on route:

![image](https://user-images.githubusercontent.com/1386719/74615502-67756300-5119-11ea-93d0-dfaf58627646.png)

And so on?

aubergine10 commented:
All 2-lane 2-way roads:

![image](https://user-images.githubusercontent.com/1386719/74615554-d652bc00-5119-11ea-8bda-aff1208aa3d8.png)

aubergine10 commented:
Green van is entering via 1-lane 1-way road (other roads are 2-lane 2-way) but has no way to access the blue node road:

![image](https://user-images.githubusercontent.com/1386719/74615585-27fb4680-511a-11ea-99da-ea345abcd535.png)

aubergine10 commented:
3 lane asymmetric at top, 4-lane 2-way either side:

![image](https://user-images.githubusercontent.com/1386719/74615633-86c0c000-511a-11ea-95ee-eb9e7dbc7b82.png)

aubergine10 commented:
![image](https://user-images.githubusercontent.com/1386719/74615644-a1933480-511a-11ea-906f-d1169129ffaf.png)

aubergine10 commented:
Motorway transition between 4-lane 2-way and 6-lane 2-way:

aubergine10 commented:
This one should have worked by lane arithmetic alone:

![image](https://user-images.githubusercontent.com/1386719/74615703-fc2c9080-511a-11ea-8d1b-7b35da198491.png)

aubergine10 commented:
I suspect new users will do what I just did and dive in trying to use quick setup on all sorts of junctions. 

Maybe it needs some feedback (red node circle appears briefly?) to let users know when it can't do anyting with a node?

> stay in lane no longer does anything if the junction lanes do not match. Is that OK?

I would suggest trying to make it deal with +/- 1 lane for given direction. E.g. 2 lanes merge in to 1, or 1 lane diverges in to 2 lanes. Seems to be a fairly common thing where road widens or narrows (gains or loses lanes) where user would want only the outer lanes to be affected.

kianzarrin commented:
> Maybe it needs some feedback (red node circle appears briefly?) to let users know when it can't do anything with a node?

I can add a guide to explain it. my worst nightmare: That requires new Crowdin keys!

kianzarrin commented:
This is more like a case of very limited allocation of dedicated lanes. so maybe I should call it that.

kvakvs commented:
@kianzarrin maybe you have an idea how to give visual feedback when auto-allocation is possible (supported). Like render a contrast node overlay or something. We don't have an easy way to show localized text over the world at this moment, only debug text renders which are not usable.

kianzarrin commented:
> @kvakvs  maybe you have an idea how to give visual feedback when auto-allocation is possible (supported). Like render a contrast node overlay or something. We don't have an easy way to show localized text over the world at this moment, only debug text renders which are not usable.

We have guide remember. I can also add visual feedback.

@aubergine10 I am waiting for you to agree with the scope of the review before I upload the second round.

kianzarrin commented:
@aubergine10 @kvakvs  Providing visual feedback that shift+S (dedicated lane connections) is possible gives users the wrong impression that its useful in that particular circumstance. Just because you can press shift+S does not mean that you should. 

There was a issue regarding showing possible keybinds at the bottom of screen. maybe that is a good place to address this?

kvakvs commented:
For keybinds i am planning to either have a panel docked at top or bottom side of TMPE main panel, or to extend TMPE main panel by a couple text rows and fit text there. It does not solve "availability visual feedback", but it will help user awareness of keybinds available.

aubergine10 commented:
I agree with the scope of the review; just need to make sure that by the time these updates reach the user that there's some indication and/or explaination where it won't set things up (eg. 4-way junctions). It doesn't need to be much, but just some confirmation to user that "yes, I noticed your interaction but won't do anything for this junction". A momentry red node circle would suffice initially.

In situations where it's only going to add connectors to a few lanes, but not all lanes, it might be better to not do anything (ie. red node circle when user tries the shortcut). It feels odd when a junction is only partly lane connected, like something broke (from end-user perspective).

> > Also, still seems to be using Shift + S rather than Ctrl + S?
  
> Its configurable in the settings. Do you wan to change the default settings? at some point it was shift+S then shift+D and now ctrl+S . we really need to make up our minds!

I think default should be `Ctrl + <some key>` (as `Shift` is now firmly rooted as 'apply to route' across all tools). I'm not particularly fussed about the choice of `<some key>` although I'd suggest something in close proximity to `Ctrl` key on the keyboard.

kianzarrin commented:
OK I will connect the oppsite side too (state machines need 4 states: forward/backward/both/none).
Guide message for when control+S is not supported
CTRL+S instead of SHIFT+S

kianzarrin commented:
@aubergine10 CTRL+S conflicts with modtools.

kianzarrin commented:
test pics:
![Screenshot (601)](https://user-images.githubusercontent.com/26344691/75398331-0fe9ab00-5902-11ea-81a3-8d38d0f8ec71.png)
![Screenshot (603)](https://user-images.githubusercontent.com/26344691/75398337-14ae5f00-5902-11ea-80cf-0daf0966dd4c.png)
![Screenshot (602)](https://user-images.githubusercontent.com/26344691/75398341-1841e600-5902-11ea-99a7-e73b1de8ce77.png)

aubergine10 commented:
Can we expose the Ctrl+S shortcut in mod options Keybinds tab? So user can change it if they run in to problems?

_Note: While there is potentially an alternative using modifier + click, that would cause problems if we at some point in the future merge lane arrows and lane connectors tools (with lane routing tool active: click segment to do arrows, node to do connectors)._

kianzarrin commented:
> @aubergine10 Can we expose the Ctrl+S shortcut in mod options Keybinds tab? So user can change it if they run in to problems?

this is already the case

kianzarrin commented:
Eh what the heck I just did 4 way stay in lane too.

also removed the restriction that lane arithmetic shall work (but if your lanes are ridiculous then expect ridiculous  outcome!).

The length of the code has been shortened a lot by this generalization.

I also removed the old stay in lane function because my code is basically a more  general form of stay in lane.
![Screenshot (611)](https://user-images.githubusercontent.com/26344691/75498816-de8be080-59d0-11ea-84bb-969785d77ac8.png)
![Screenshot (612)](https://user-images.githubusercontent.com/26344691/75498812-dc298680-59d0-11ea-8eb8-11f5dc90986f.png)
![Screenshot (613)](https://user-images.githubusercontent.com/26344691/75498808-d9c72c80-59d0-11ea-8fe9-b3085d5a1ef2.png)
![Screenshot (614)](https://user-images.githubusercontent.com/26344691/75498803-d6cc3c00-59d0-11ea-9f24-c87955beb76c.png)

![Screenshot (617)](https://user-images.githubusercontent.com/26344691/75498873-0d09bb80-59d1-11ea-94a5-24e950c432c2.png)

aubergine10 commented:
Looking good!! 👍 I'll test this later.

aubergine10 commented:
Possible to make it handle number of lanes changing by +/-1 a bit better?

![image](https://user-images.githubusercontent.com/1386719/75714831-52c1cf00-5cc4-11ea-8c6d-286e2f504404.png)

![image](https://user-images.githubusercontent.com/1386719/75715289-1478df80-5cc5-11ea-8a09-90f8193147bd.png)

And this would prevent access to one of the lanes on the 2-lane 1-way road:

![image](https://user-images.githubusercontent.com/1386719/75715598-91a45480-5cc5-11ea-9242-cb21680b14a2.png)

And would it be possible to handle this a bit better (red and green go straight on, blue collides with red)?

![image](https://user-images.githubusercontent.com/1386719/75715746-d29c6900-5cc5-11ea-9036-4bbcc4b59b1c.png)

Other than that it's working really well. Using the shortcut multiple times does one side, then the other, then back to both sides.

kianzarrin commented:
@aubergine10 
> would it be possible to handle this a bit better (red and green go straight on, blue collides with red)?
![image](https://user-images.githubusercontent.com/1386719/75715746-d29c6900-5cc5-11ea-9036-4bbcc4b59b1c.png)
well you say that because you know it is  a roundabout. if this was a highway you would like to do it like bellow ... 
![Screenshot (646)](https://user-images.githubusercontent.com/26344691/75720967-1c924880-5ce0-11ea-9ece-213214b82804.png)
As opposed to 
![Screenshot (648)](https://user-images.githubusercontent.com/26344691/75720988-261bb080-5ce0-11ea-99ac-fa374ded70e7.png)

am I wrong?

kianzarrin commented:
> Possible to make it handle number of lanes changing by +/-1 a bit better?
![image](https://user-images.githubusercontent.com/1386719/75714831-52c1cf00-5cc4-11ea-8c6d-286e2f504404.png)
![image](https://user-images.githubusercontent.com/1386719/75715289-1478df80-5cc5-11ea-8a09-90f8193147bd.png)

remember when I said in the description:
> the only functional change to stay in lane is when lane arithmetic does not match, outer lanes are connected first rather than inner lanes. I consider this to be improvement.

I should change it back to what it was.

kianzarrin commented:
> this would prevent access to one of the lanes on the 2-lane 1-way road:
![image](https://user-images.githubusercontent.com/1386719/75715598-91a45480-5cc5-11ea-9242-cb21680b14a2.png)

my intention was if the lane arithmetic does not work the user should connect the final lane manually. I will add code to connect the one final lane if target - source == 1. 

what if the difference is +-2? this can easily get out of hand.

maybe you are thinking this does not fit into our plan to introduce intent ... but with intent we are talking about preferred direction rather than forced direction so i think it will work out.

kianzarrin commented:
@aubergine10 User should not use lane connector stay-in-lane for roundabouts (that is my intention). use lane-arrows allocated dedicated turning lanes instead.

kianzarrin commented:
I made the code better little by little over the past week. Next round of review is ready
@aubergine10 Please don't request extras features. this review has been dragged on long enough.
@kvakvs I tried to provide as many layers of abstraction as I could and combined with a lot of calculations, I hope it is understandable.

kianzarrin commented:
Appveyer faild. I don't understand why:
```
  UI\SubTools\LaneConnectorTool.cs(34,48): error CS0103: The name 'DebugSwitch' does not exist in the current context 
```
It builds fine on my computer.

kvakvs commented on TLM/TLM/Util/Shortcuts.cs:
```diff
@@ -19,17 +17,35 @@ internal static class Shortcuts {
         internal static IList<T> Clone<T>(this IList<T> listToClone) where T : ICloneable =>
             listToClone.Select(item => (T)item.Clone()).ToList();
 
+        internal static void Swap<T>(ref T a, ref T b) {
```
C# 7 introduced tuples which enables swapping two variables without a temporary one:
```
int a = 10;
int b = 2;
(a, b) = (b, a);
```

kianzarrin replied:
Yeah I know! but this is not c# 7 :(

kvakvs replied:
We have all C# versions available. Only .NET framework is limited to 3.5.

krzychu124 replied:
Actually... not. Some of the features still need additional namespaces not available for .net 3.5, sadly :/

dymanoid replied:
@kvakvs Well, in fact, there will be even **two** temporary variables with your solution. Check the generated IL. And @krzychu124 is right: the .NET Framework 3.5 BCL has no value tuples, as well as Mono of Unity v5, so this is not possible without an additional NuGet package (which I wouldn't recommend).

kvakvs commented on TLM/TMPE.CitiesGameBridge/TMPE.CitiesGameBridge.csproj:
```diff
@@ -1,4 +1,4 @@
-<?xml version="1.0" encoding="utf-8"?>
+﻿<?xml version="1.0" encoding="utf-8"?>
```
What exactly changed here, is it BOM?

kianzarrin replied:
Yeah. I tried to add a new file but then changed my mind and deleted it. that's why the bom has changed.

kianzarrin commented:
two way road connected from inner side is not supported see https://github.com/CitiesSkylinesMods/TMPE/issues/788#issuecomment-599812119

kianzarrin commented:
> @aubergine10 please rebase this PR to master for SH compatibility
DONE

aubergine10 commented:
It's working really great now.

<details><summary>Minor issue with 3-lane asymetric roads:</summary>
Some lanes left unconnected:

![image](https://user-images.githubusercontent.com/1386719/78329381-76a15a80-7579-11ea-8094-5022220ea9f6.png)

Also, doing Ctrl+S again had no effect; could we always have a state that is "no connections" as that would let user know more clearly that they've cycled through all available options.</details>

<details><summary>Another minor issue with 3-lane asym road:</summary>
1. Ctrl+S gives this:

![image](https://user-images.githubusercontent.com/1386719/78329587-0515dc00-757a-11ea-8794-a1fe0e411f24.png)

Then:

2.  second Ctrl+S does nothing
3. Ctrl+S removes the connectors
4. Ctrl+S does nothing
5. Ctrl+S goes back to start (image above)
    * etc...</details>

<details><summary>stay-in-lane not working on highway?</summary>
Ctrl+S had no effect here:

![image](https://user-images.githubusercontent.com/1386719/78329759-69d13680-757a-11ea-8f3c-deab25f48585.png)

I would expect:

1. Ctrl+S = Stay in lane (all lanes connected ahead only)
2. another Ctrl+S = ahead + adjacent lanes
3. another Ctrl+S = back to no connectors
</details>

kianzarrin commented:
> Some lanes left unconnected:
![image](https://user-images.githubusercontent.com/1386719/78329381-76a15a80-7579-11ea-8094-5022220ea9f6.png)
>Another minor issue with 3-lane asym road

@aubergine10 Because it only has one lane! Do you remember you were complaining roads that have only one lane should not use stay in lane? well you got your wish!!!!

kianzarrin commented:
@aubergine10 
> stay-in-lane not working on highway?

Nice catch. It was intermittent bug with an easy fix.

kianzarrin commented:
> @aubergine10 I would expect ... another Ctrl+S = ahead + adjacent lanes

Not in this PR. for the next phase  (#788) its a good idea.

aubergine10 commented:
https://github.com/CitiesSkylinesMods/TMPE/pull/706#issuecomment-608986884

> Because it only has one lane! Do you remember you were complaining roads that have only one lane should not use stay in lane? well you got your wish!!!!

Yet it's adding connectors in the other direction, which is essentially the same situation ;)

aubergine10 commented:
It's working great now!

Regarding https://github.com/CitiesSkylinesMods/TMPE/pull/706#issuecomment-608986884 (no connector for single lane), I think the rule should be:

> _Don't add a connector if the source segment only has single lane entering the node_

So if a road has 2+ lanes entering a node, it should be ok to connect those up to target lane(s), even if target segment only has single lane leaving that node.

What do you think?

aubergine10 commented:
^ edited comment above

kianzarrin commented:
> @aubergine10  I think the rule should be: Skip the "don't connect single lanes" rule at junction/transition nodes

Agreed. Is that the last issue you could find.

kianzarrin commented:
@aubergine10 fixed the one lane road issue.

aubergine10 commented:
<details><summary>Soooo close</summary>
Shown here the result of Ctrl+S and where I was hoping red connector would go:

![image](https://user-images.githubusercontent.com/1386719/78613758-8ca66800-7864-11ea-95aa-508ea9d5e1d6.png)

It's doing the road entering from bottom correctly, but the road entering from top (which is essentially same situation as bottom road) isn't connecting to the single-lane 1-way street existing to the left.

If it could somehow detect that it would be awesome, but overall I think this PR is pretty much complete.</details>

kianzarrin commented:
sorry request review by mistake!
